File: Sources/Services/PendingLeaveStore.swift
import Foundation

/// Tracks groups where the user has requested to leave but the admin
/// hasn't yet processed the removal (which triggers MLS key rotation).
///
/// UserDefaults-backed so the "Leaving…" state survives app restarts.
@MainActor
final class PendingLeaveStore: ObservableObject {

    private static let storageKey = "fmf.pendingLeaves"

    @Published private(set) var pendingLeaves: Set<String> = []  // groupIds

    init(skipLoad: Bool = false) {
        if !skipLoad { load() }
    }

    // MARK: - Queries

    func contains(_ groupId: String) -> Bool {
        pendingLeaves.contains(groupId)
    }

    // MARK: - Mutations

    func add(_ groupId: String) {
        guard !pendingLeaves.contains(groupId) else { return }
        pendingLeaves.insert(groupId)
        save()
        FMFLogger.marmot.info("PendingLeaveStore: added leave request for group \(groupId)")
    }

    func remove(_ groupId: String) {
        guard pendingLeaves.contains(groupId) else { return }
        pendingLeaves.remove(groupId)
        save()
        FMFLogger.marmot.info("PendingLeaveStore: removed leave request for group \(groupId)")
    }

    /// Remove all pending leaves (used during identity replacement).
    func removeAll() {
        pendingLeaves.removeAll()
        save()
    }

    /// Remove any pending leaves for groups that no longer exist in the active
    /// group list — meaning the admin processed the removal successfully.
    func removeResolved(activeGroupIds: Set<String>) {
        let resolved = pendingLeaves.subtracting(activeGroupIds)
        guard !resolved.isEmpty else { return }
        pendingLeaves.subtract(resolved)
        save()
        FMFLogger.marmot.info("PendingLeaveStore: auto-cleared \(resolved.count) resolved leave(s)")
    }

    // MARK: - Persistence

    private func save() {
        guard let data = try? JSONEncoder().encode(Array(pendingLeaves)) else { return }
        UserDefaults.standard.set(data, forKey: Self.storageKey)
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: Self.storageKey),
              let decoded = try? JSONDecoder().decode([String].self, from: data) else {
            return
        }
        pendingLeaves = Set(decoded)
    }
}

File: android/app/src/main/java/org/findmyfam/services/PendingLeaveStore.kt
package org.findmyfam.services

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Tracks groups where the user has requested to leave but the admin
 * hasn't yet processed the removal (which triggers MLS key rotation).
 *
 * SharedPreferences-backed so the "Leaving..." state survives app restarts.
 */
@Singleton
class PendingLeaveStore @Inject constructor(
    @ApplicationContext context: Context
) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences("fmf_pending_leaves", Context.MODE_PRIVATE)

    private val _pendingLeaves = MutableStateFlow<Set<String>>(emptySet())
    val pendingLeaves: StateFlow<Set<String>> = _pendingLeaves.asStateFlow()

    init {
        load()
    }

    fun contains(groupId: String): Boolean {
        return groupId in _pendingLeaves.value
    }

    fun add(groupId: String) {
        if (groupId in _pendingLeaves.value) return
        _pendingLeaves.value = _pendingLeaves.value + groupId
        save()
        Timber.i("PendingLeaveStore: added leave request for group $groupId")
    }

    fun remove(groupId: String) {
        if (groupId !in _pendingLeaves.value) return
        _pendingLeaves.value = _pendingLeaves.value - groupId
        save()
        Timber.i("PendingLeaveStore: removed leave request for group $groupId")
    }

    fun removeAll() {
        _pendingLeaves.value = emptySet()
        save()
    }

    /**
     * Remove any pending leaves for groups that no longer exist in the active
     * group list -- meaning the admin processed the removal successfully.
     */
    fun removeResolved(activeGroupIds: Set<String>) {
        val resolved = _pendingLeaves.value - activeGroupIds
        if (resolved.isEmpty()) return
        _pendingLeaves.value = _pendingLeaves.value - resolved
        save()
        Timber.i("PendingLeaveStore: auto-cleared ${resolved.size} resolved leave(s)")
    }

    // --- Persistence ---

    private fun save() {
        val arr = JSONArray()
        for (id in _pendingLeaves.value) arr.put(id)
        prefs.edit().putString(STORAGE_KEY, arr.toString()).apply()
    }

    private fun load() {
        val json = prefs.getString(STORAGE_KEY, null) ?: return
        try {
            val arr = JSONArray(json)
            val set = mutableSetOf<String>()
            for (i in 0 until arr.length()) set.add(arr.getString(i))
            _pendingLeaves.value = set
        } catch (e: Exception) {
            Timber.w(e, "Failed to load pending leaves")
        }
    }

    companion object {
        private const val STORAGE_KEY = "fmf.pendingLeaves"
    }
}
